test/integration/otel_test.go

```diff
@@ -0,0 +1,136 @@
+//go:build integration
+
+package integration
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"go.opentelemetry.io/otel/sdk/trace"
+
+	"github.com/letsencrypt/boulder/test"
+)
+
+// TraceResponse is just the minimal fields we care about from the Jaeger traces API
+type TraceResponse struct {
+	Data []struct {
+		Spans []struct {
+			OperationName string
+		}
+	}
+}
+
+func getWFEOperations(t *testing.T, op string) *TraceResponse {
+	tracesURL := "http://jaeger:16686/api/traces?service=wfe2&operation=" + url.QueryEscape(op)
+	fmt.Println(tracesURL)
+	resp, err := http.Get(tracesURL)
+	test.AssertNotError(t, err, "failed to fetch operation from jaeger")
+	test.AssertEquals(t, resp.StatusCode, http.StatusOK)
+
+	body, err := io.ReadAll(resp.Body)
+	test.AssertNotError(t, err, "failed to read traces body")
+
+	var parsed TraceResponse
+	err = json.Unmarshal(body, &parsed)
+	test.AssertNotError(t, err, "failed to decode traces body")
+
+	test.Assert(t, len(parsed.Data) > 0, fmt.Sprintf("expected to get traces: %v (%s)", parsed, string(body)))
+
+	return &parsed
+}
+
+// assertSpans checks that all the passed in traces contain the list of spans
+func assertSpans(t *testing.T, tr *TraceResponse, spans []string) {
+	for _, trace := range tr.Data {
+		for _, span := range spans {
+			found := false
+			for _, a := range trace.Spans {
+				if span == a.OperationName {
+					found = true
+					break
+				}
+			}
+			test.Assert(t, found, fmt.Sprintf("Didn't find expected span: %s in %v", span, trace))
+		}
+	}
```

Comment:
I would expect this loop to fail -- many of our integration tests purposefully fail issuance for one reason or another, so I wouldn't expect (for example) every /acme/finalize/ trace to actually contain a ca.Certificateauthority/IssuePrecertificate span. Similarly I wouldn't expect every /acme/new-order trace to contain a sa.StorageAuthority/NewOrderAndAuthzs span, since some of those requests are designed to fail due to rate limits or invalid requested names.

I think this is going to need some way to filter for only the traces created by this individual integration test, perhaps by filtering down to those that have the correct set of domains set in one of their fields?

Reply:
I think we probably want to set a trace ID on the requests generated in the test, and then search for those traces specifically.  This test is definitely jank and not ready yet.

Reply:
Pushed a WIP update:  It seems like making the entire test run be a single trace will work out well.  Had to be a bit hacky with the ACME client as it doesn't pass contexts through to the HTTP client.

Reply:
I pushed a less-hacky test that wraps the test run in a single trace, which we can pull out of Jaeger and verify all the expected spans are present with the right structure.